test/parallel/test-fs-rm.js

```diff
@@ -287,74 +286,154 @@ function removeAsync(dir) {
-        } else {
-          fs.chmodSync(dir, 0o777);
+      {
+        const dirname = nextDirPath();
+        const filePath = path.join(dirname, 'text.txt');
+
+        fs.mkdirSync(dirname, { recursive: true });
+        fs.writeFileSync(filePath, 'hello');
+
+        fs.chmodSync(filePath, 0o444);
+        fs.chmodSync(dirname, 0o444);
+
+        let err = null;
+
+        try {
+          fs.rmSync(filePath, { force: true });
+        } catch (_err) {
+          err = _err;
+        }
+
+        try {
+          fs.chmodSync(dirname, 0o777);
+        } catch {
+        }
+
+        try {
+          fs.chmodSync(filePath, 0o777);
+        } catch {
+        }
+
+        if (!isValidState(fs.existsSync(filePath), err)) {
+          throw err;
         }
       }
-    }
 
-    {
-      // Check that deleting a file that cannot be accessed using rmsync throws
-      // https://github.com/nodejs/node/issues/38683
-      const dirname = nextDirPath();
-      const filePath = path.join(dirname, 'text.txt');
-      try {
+      {
+        const dirname = nextDirPath();
+        const filePath = path.join(dirname, 'text.txt');
+
         fs.mkdirSync(dirname, { recursive: true });
         fs.writeFileSync(filePath, 'hello');
-        const code = makeDirectoryReadOnly(dirname, 0o444);
-        assert.throws(() => {
-          fs.rmSync(filePath, { force: true });
-        }, {
-          code,
-          name: 'Error',
-        });
-      } finally {
-        makeDirectoryWritable(dirname);
+
+        fs.chmodSync(filePath, 0o444);
+        fs.chmodSync(dirname, 0o444);
+
+        fs.rm(filePath, { force: true }, common.mustCall((err) => {
+          try {
+            fs.chmodSync(dirname, 0o777);
+          } catch {
+          }
+
+          try {
+            fs.chmodSync(filePath, 0o777);
+          } catch {
+          }
+
+          if (!isValidState(fs.existsSync(filePath), err)) {
+            throw err;
+          }
+        }));
       }
     }
 
+    // Check endless recursion.
+    // https://github.com/nodejs/node/issues/34580
     {
-      // Check endless recursion.
-      // https://github.com/nodejs/node/issues/34580
-      const dirname = nextDirPath();
-      fs.mkdirSync(dirname, { recursive: true });
-      const root = fs.mkdtempSync(path.join(dirname, 'fs-'));
-      const middle = path.join(root, 'middle');
-      fs.mkdirSync(middle);
-      fs.mkdirSync(path.join(middle, 'leaf')); // Make `middle` non-empty
-      try {
-        const code = makeDirectoryReadOnly(middle, 0o555);
+      function isValidState(exists, err) {
+        // On Windows, we are not allowed to delete a read-only folder yet.
+        // TODO(RaisinTen): Remove Windows special-casing if this lands:
+        // https://github.com/libuv/libuv/pull/3193
+        return common.isWindows ?
+          (exists === true && err?.code === 'EPERM') :
+          (exists === true && err?.code === 'EACCES');
```

Comment:
nit: not sure if you notice that `exists` state is not Windows dependent, please discard if you chose to repeat it on purpose.
```suggestion
        return exists === true &&
               err?.code === common.isWindows ? 'EPERM' : 'EACCES';
```